core/modules/taxonomy/taxonomy.module

```diff
@@ -1852,6 +1857,26 @@ function taxonomy_field_widget_form(&$form, &$form_state, $field, $instance, $la
   return $element;
 }
 
+/**
+ * Implements hook_field_widget_options_buttons_form_alter().
+ */
+function taxonomy_field_widget_options_buttons_form_alter(&$element, &$form_state, $context) {
+  if ($context['field']['type'] == 'taxonomy_term_reference') {
+    $options = taxonomy_allowed_values($context['field'], FALSE);
+    foreach ($context['field']['settings']['allowed_values'] as $tree) {
```

Comment:
This line assumes that the field info will always have `'allowed_values'` set. This is not always the case. It's possible to use `$field['settings']['options_list_callback']` instead, to specify a callback that's different from the default `taxonomy_allowed_values` to obtain the options.

Reply:
I've added a check for the existence of `$context['field']['settings']['allowed_values']`, and if it's not present, added code to rebuild the essential bits of it. It seems a bit heavy (fuly loading all the taxonomy terms to get their vocabularies), and I'm thinking it might be doable with a single direct db query; so looking into that.